File: multifactor/sp_vs_gs10_naive_bayes_dataprep.R
# A Naive Bayes model for market timing
# Investing in the S&P 500 or 10-year treasury based on various factors

# dataprep only: join all relevant datasets and perform all feature engineering

library(tidyverse)
library(lubridate)
library(e1071)
library(caret)
library(reshape2)

# S&P 500 and 10-year treasury returns data
# from monthly since 1871
sp.data <- read.csv("./prepared_data/schiller_sp_data_withreturns.tsv", sep="\t", stringsAsFactors = FALSE)
head(sp.data)
tail(sp.data)
sp.data$Date <- as.Date(sp.data$Date, format="%Y-%m-%d")

train.test.cutoff <- as.Date("2005-01-01")

## Compute and/or join with predictors

# Add Momentum Factors
sp.data$SP.Momentum.6Mo <- sp.data$SP.Price - sp.data$SP.Price.Prev.6Mo
sp.data$SP.Momentum.6Mo.Negative <- sp.data$SP.Momentum.6Mo < 0
sp.data$SP.Momentum.1Mo <- sp.data$SP.Price - sp.data$SP.Price.Last
sp.data$SP.Momentum.1Mo.Negative <- sp.data$SP.Momentum.1Mo < 0
sp.data$SP.Momentum.12Mo <- sp.data$SP.Price - sp.data$SP.Price.Prev.12Mo
sp.data$SP.Momentum.12Mo.Negative <- sp.data$SP.Momentum.12Mo < 0

# Add Yield Curve Status
yield.curve.data <- read.csv("./prepared_data/yield_curve_10y_3mo.tsv",sep="\t")
head(yield.curve.data)
yield.curve.data$DATE <- as.Date(yield.curve.data$DATE, format="%Y-%m-%d")
# Only keep needed columns
yield.curve.data <- yield.curve.data %>% select(DATE,GS10.Tbill.Spread,Yield.Curve.Status)
sp.data <- sp.data %>% left_join(yield.curve.data, by=c("Date"="DATE"))
table(sp.data$Yield.Curve.Status)
sp.data$Yield.Curve.Inverted <- sp.data$Yield.Curve.Status=="Inverted"

# Add AAII Sentiment Data
sentiment.data <- read.csv("./prepared_data/sentiment_data_formatted.tsv", sep="\t", stringsAsFactors = FALSE)
head(sentiment.data)
sentiment.data$ReportedDate <- lubridate::mdy(sentiment.data$ReportedDate)
# Associate the most recent sentiment data report with each month
sentiment.data$CeilingMonth <- sentiment.data$ReportedDate %>%
  ceiling_date(unit = "months")
sentiment.data.monthly <- sentiment.data %>%
  arrange(CeilingMonth, desc(ReportedDate))
sentiment.data.monthly <- sentiment.data.monthly %>%
  distinct(CeilingMonth, .keep_all = TRUE)
head(sentiment.data.monthly)
# join with the S&P data
sentiment.data.monthly <- sentiment.data.monthly %>% select(CeilingMonth,Bullish)
sp.data <- sp.data %>%
  left_join(sentiment.data.monthly, by=c("Date"="CeilingMonth"))
# construct a bineary feature
sp.data$Bullish.High <- sp.data$Bullish > 0.5

# Add Risk Premium vs T-bill rate
TB3MS <- read.csv("./raw_data/TB3MS.csv", stringsAsFactors = FALSE)
names(TB3MS)[2] <- "Tbill.Rate"
TB3MS$Tbill.Rate <- TB3MS$Tbill.Rate / 100
TB3MS$DATE <- as.Date(TB3MS$DATE)
head(TB3MS)
sp.data <- sp.data %>% 
  left_join(TB3MS,by=c("Date"="DATE"))

# use CAPE for earnings yield for smoothing
sp.data$Earnings.Yield <- 1/sp.data$CAPE
sp.data$SP.Risk.Premium.Tbill <- sp.data$Earnings.Yield - sp.data$Tbill.Rate

sp.data$Low.Risk.Premium <- sp.data$SP.Risk.Premium.Tbill < 0.01
table(sp.data$Low.Risk.Premium)

# Construct the target variable
# For a Naive Bayes model we need a categorical target, so just use a binary variable
# indicating whether the S&P 500 or 10-year treasury performs better 
# on a one-month forward basis
sp.data <- sp.data %>%
  mutate(
    SP.Outperforms.GS10 = as.factor(SP.Return.Forward > GS10.Return.Forward)
  )
summary(sp.data$SP.Outperforms.GS10)

File: multifactor/sp_vs_gs10_naive_bayes_dataprep.py
import pandas as pd
import numpy as np

## S&P 500 and 10-year treasury returns data monthly since 1871, from Schiller dataset
schiller_data_prepared_fname = "../prepared_data/schiller_sp_data_withreturns.tsv"
schiller_data = pd.read_csv(schiller_data_prepared_fname, sep="\t")
schiller_data.dtypes

# Calculate Momentum Factors
# use np.sign to preserve NaN when one value is unavailable
schiller_data['SP.Momentum.1Mo.Negative'] = np.sign(schiller_data['SP.Price.Last'] - schiller_data['SP.Price'])
schiller_data['SP.Momentum.6Mo.Negative'] = np.sign(schiller_data['SP.Price.Prev.6Mo'] - schiller_data['SP.Price'])
schiller_data['SP.Momentum.12Mo.Negative'] = np.sign(schiller_data['SP.Price.Prev.12Mo'] - schiller_data['SP.Price'])
# relabel sign=-1 to sign=0
schiller_data['SP.Momentum.1Mo.Negative'].mask(schiller_data['SP.Momentum.1Mo.Negative']<0,0,inplace=True)
schiller_data['SP.Momentum.6Mo.Negative'].mask(schiller_data['SP.Momentum.6Mo.Negative']<0,0,inplace=True)
schiller_data['SP.Momentum.12Mo.Negative'].mask(schiller_data['SP.Momentum.12Mo.Negative']<0,0,inplace=True)
# verify on latest 100 entries for each
(schiller_data['SP.Price.Last'] < schiller_data['SP.Price'])[-100:-1].value_counts()
schiller_data['SP.Momentum.1Mo.Negative'][-100:-1].value_counts()
(schiller_data['SP.Price.Prev.6Mo'] < schiller_data['SP.Price'])[-100:-1].value_counts()
schiller_data['SP.Momentum.6Mo.Negative'][-100:-1].value_counts()
(schiller_data['SP.Price.Prev.12Mo'] < schiller_data['SP.Price'])[-100:-1].value_counts()
schiller_data['SP.Momentum.12Mo.Negative'][-100:-1].value_counts()

## AAII Sentiment Data
aaii_filename = "../prepared_data/sentiment_data_formatted.tsv"
aaii_data = pd.read_csv(aaii_filename, sep="\t")

# Associate the most recent sentiment data report with each month
aaii_data = aaii_data.astype({"ReportedDate": np.datetime64})
aaii_data['Ceiling.Month'] = aaii_data['ReportedDate'] + pd.offsets.MonthBegin(0)
# keep only the latest entry for each month
aaii_data_reverse = aaii_data.sort_index(ascending=False)
aaii_data_reverse_dupl = aaii_data_reverse['Ceiling.Month'].duplicated()
aaii_data_monthly = aaii_data_reverse[aaii_data_reverse_dupl==False].sort_index(ascending=True)

## T-bill rate for risk premium calculation
TB3MS = pd.read_csv("../raw_data/TB3MS.csv")
TB3MS['Tbill.Rate'] = np.divide(TB3MS['TB3MS'],100)

## 10-year treasury data series for treasury spread (aka yield curve)
GS10 = pd.read_csv("../raw_data/GS10.csv")
GS10['Treasury.10Y.Rate'] = np.divide(GS10['GS10'],100)

# format dates
schiller_data.dtypes
GS10.dtypes
schiller_data = schiller_data.astype({"Date": np.datetime64})
GS10 = GS10.astype({"DATE": np.datetime64})
TB3MS = TB3MS.astype({"DATE": np.datetime64})

## join all the tables

# join 10-year treasury data
joined_data = pd.merge(schiller_data, GS10, how="left", left_on="Date", right_on="DATE")
# verify the join
schiller_data.shape
joined_data.shape
schiller_data['Date'].tail()
GS10['DATE'].head()
GS10['DATE'].tail()
joined_data[['Date','Treasury.10Y.Rate']].head()
joined_data[['Date','Treasury.10Y.Rate']].tail()
GS10['Treasury.10Y.Rate'].describe()
joined_data['Treasury.10Y.Rate'].describe()

# add t-bill rate data
joined_data = pd.merge(joined_data, TB3MS, how="left", left_on="Date", right_on="DATE")
joined_data.shape
joined_data[['Date','TB3MS']].head()
joined_data[['Date','TB3MS']].tail()
joined_data['TB3MS'].describe()
TB3MS['TB3MS'].describe()
joined_data['Tbill.Rate'].tail()

# add aaii data
aaii_data_monthly.head()
joined_data = pd.merge(joined_data, aaii_data_monthly, how="left", left_on="Date", right_on="Ceiling.Month")
joined_data.shape
joined_data[['Date','Bullish']].head()
joined_data[['Date','Bullish']].tail()
aaii_data_monthly['Bullish'].describe()
joined_data['Bullish'].describe()

# TODO calculate yield curve
# import pdb; pdb.set_trace()
joined_data['GS10.Tbill.Spread'] = np.subtract(joined_data['Treasury.10Y.Rate'], joined_data['Tbill.Rate'])
joined_data['GS10.Tbill.Spread'].describe()
joined_data['GS10.Tbill.Spread.Positive'] = np.sign(joined_data['GS10.Tbill.Spread'])
joined_data['GS10.Tbill.Spread.Positive'].value_counts()
# consider 0 spread to be negative
joined_data['GS10.Tbill.Spread.Positive'].mask(joined_data['GS10.Tbill.Spread.Positive']<0,0,inplace=True)
joined_data['GS10.Tbill.Spread.Positive'].value_counts()

# calculate risk premium
# use CAPE for earnings yield for smoothing
joined_data['Earnings.Yield'] = np.divide(1,joined_data['CAPE'])
joined_data[['Date','Earnings.Yield','CAPE']].tail()
joined_data['SP.Risk.Premium.Tbill'] = np.subtract(joined_data['Earnings.Yield'], joined_data['Tbill.Rate'])
joined_data[['Date','SP.Risk.Premium.Tbill','Earnings.Yield','Tbill.Rate']].tail()
# use np.sign to preserve NaN cases
joined_data['Low.Risk.Premium'] = np.sign(np.subtract(0.01,joined_data['SP.Risk.Premium.Tbill']))
# consider cases where risk premium = 0.01 to be high risk premium
joined_data['Low.Risk.Premium'].mask(joined_data['Low.Risk.Premium']<0,0,inplace=True)

# Construct the target variable
# For a Naive Bayes model we need a categorical target, so just use a binary variable
# indicating whether the S&P 500 or 10-year treasury performs better 
# on a one-month forward basis

# use np.sign to preserve NaN cases
joined_data['SP.Outperforms.GS10'] = np.sign(np.subtract(joined_data['SP.Return.Forward'], joined_data['GS10.Return.Forward']))
# tie goes to bonds
joined_data['SP.Outperforms.GS10'].mask(joined_data['SP.Outperforms.GS10']<0,0,inplace=True)

# verify
joined_data[['Date','SP.Outperforms.GS10','SP.Return.Forward','GS10.Return.Forward']].tail()
joined_data['SP.Outperforms.GS10'].value_counts()

joined_data.to_csv('./naive_bayes_model_training_data.tsv',sep="\t",index=False)

